onnxruntime/test/python/onnxruntime_test_collective.py

```diff
@@ -154,42 +155,50 @@ def _create_alltoall_ut_model_for_boolean_tensor(
-                model.SerializeToString(),
-                providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
-                provider_options=[{"device_id": str(rank)}, {}],
-            )
+    @parameterized.expand(
+        [
+            (np.float32, TensorProto.FLOAT),
+        ]
+    )
+    def test_all_reduce(self, np_elem_type, elem_type):
+        model = self._create_allreduce_ut_model((128, 128), elem_type)
+        rank, size = self._get_rank_size()
+        ort_sess = ort.InferenceSession(
+            model.SerializeToString(),
+            providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
+            provider_options=[{"device_id": str(rank)}, {}],
+        )
 
-            input = np.ones((128, 128), dtype=np_elem_type)
-            outputs = ort_sess.run(None, {"X": input})
-            assert np.allclose(outputs[0], size * input)
-
-    def test_all_gather(self):
-        for np_elem_type, elem_type, communication_elem_type in ((np.float32, TensorProto.FLOAT, TensorProto.FLOAT),):
-            model = self._create_allgather_ut_model((128, 128), 0, elem_type, communication_elem_type)
-            rank, size = self._get_rank_size()
-            ort_sess = ort.InferenceSession(
-                model.SerializeToString(),
-                providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
-                provider_options=[{"device_id": str(rank)}, {}],
-            )
+        input = np.ones((128, 128), dtype=np_elem_type)
+        outputs = ort_sess.run(None, {"X": input})
+        assert np.allclose(outputs[0], size * input)
 
-            input = np.ones((128, 128), dtype=np.float32) * rank
-            outputs = ort_sess.run(None, {"X": input})
+    @parameterized.expand(
+        [
+            (np.float32, TensorProto.FLOAT, TensorProto.FLOAT),
+        ]
+    )
+    def test_all_gather(self, np_elem_type, elem_type, communication_elem_type):
+        model = self._create_allgather_ut_model((128, 128), 0, elem_type, communication_elem_type)
+        rank, size = self._get_rank_size()
+        ort_sess = ort.InferenceSession(
+            model.SerializeToString(),
+            providers=["CUDAExecutionProvider", "CPUExecutionProvider"],
+            provider_options=[{"device_id": str(rank)}, {}],
+        )
 
-            expected_output = np.zeros((128, 128), dtype=np_elem_type)
-            for _ in range(size - 1):
-                expected_output = np.concatenate((expected_output, np.ones((128, 128), dtype=np_elem_type) * (_ + 1)))
-            np.testing.assert_allclose(outputs[0], expected_output, err_msg="all gather on axis0: result mismatch")
+        input = np.ones((128, 128), dtype=np.float32) * rank
+        outputs = ort_sess.run(None, {"X": input})
+
+        expected_output = np.zeros((128, 128), dtype=np_elem_type)
+        for _ in range(size - 1):
+            expected_output = np.concatenate((expected_output, np.ones((128, 128), dtype=np_elem_type) * (_ + 1)))
```

Comment:
move this logic into
```
def all_gather_implementation(tensor_on_device_0, tensor_on_device_1, ...):
  ...
```

or use a more generic signature
```
def all_gather_implementation(*tensors_on_devices):
  ...
```

Reply:
Splitting this logic will make reading the test implementation harder. Moreover, the logic is only building data for a single rank so a function that takes data for all ranks as input doesn't make sense.

Adding documentation on input/output per rank to elaborate on the logic.

```
        # Input for size == 2
        # rank = 0:
        #   Input:  [[0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0], [0, 0, 0, 0]]
        #   Output: [[0, 0, 0, 0], [0, 0, 0, 0], [1, 1, 1, 1], [1, 1, 1, 1]]
        #
        # rank = 1:
        #   Input:  [[1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1], [1, 1, 1, 1]]
        #   Output: [[0, 0, 0, 0], [0, 0, 0, 0], [1, 1, 1, 1], [1, 1, 1, 1]]
```